matsim/src/main/java/org/matsim/core/network/DisallowedNextLinks.java

```diff
@@ -0,0 +1,105 @@
+package org.matsim.core.network;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import org.matsim.api.core.v01.Id;
+import org.matsim.api.core.v01.network.Link;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Class to store disallowed next links, e.g. to model turn restrictions.
+ * 
+ * @author hrewald, mrieser
+ */
+public class DisallowedNextLinks {
+
+	// Actually, it could be Map<String, Set<List<Id<Link>>>>, as the order of the
+	// next links sequences does not matter. However, we choose to store them in a
+	// list in favor of a smaller memory footprint.
+	private final Map<String, List<List<Id<Link>>>> linkIdSequencesMap = new HashMap<>();
+
+	/**
+	 * Add a sequence of subsequent links to be disallowed from the current link.
+	 * 
+	 * @param mode
+	 * @param linkSequence sequence of links that shall not be passed after passing
+	 *                     the link where this object is attached
+	 * @return true, if linkSequence was actually added
+	 */
+	public boolean addDisallowedLinkSequence(String mode, List<Id<Link>> linkSequence) {
+		List<List<Id<Link>>> linkSequences = this.linkIdSequencesMap.computeIfAbsent(mode, m -> new ArrayList<>());
+
+		boolean result = false;
+		// prevent adding empty/duplicate link id sequences, or duplicate link ids
+		if (!linkSequence.isEmpty() && new HashSet<>(linkSequence).size() == linkSequence.size()
+				&& !linkSequences.contains(linkSequence)) {
+			result = linkSequences.add(ImmutableList.copyOf(linkSequence));
+			if (result) { // sorting is required for a.equals(b) <=> a.hashCode() == b.hashCode()
```

Comment:
I am not sure I understand this comment. Did you mean: sorting is required because we use a `List` instead of a `Set` for storing all link sequences and we want to easily compare them?

Reply:
Thanks for confirming my thinking.

I think this part of the comment is confusing `a.equals(b) <=> a.hashCode() == b.hashCode()`. Instead you could add part of your response as a comment.

Reply:
Sure, just added it.